### DISCUSSION TITLE
Object type libraries?

### DISCUSSION THREAD

**Discussion by cprieto** [mesonbuild/meson#11591] [Q&A]
in CMake is possible to define _object type libraries_ (see [object libraries](https://cmake.org/cmake/help/latest/command/add_library.html#object-libraries)). Is it possible to do the same in Meson? I know I can generate a library _and then_ extract the required object to do the linking (see [object files](https://mesonbuild.com/Build-targets.html#object-files) in doc) but I was wondering, is possible to just compile the object file without making a library?

**Comment by HaoZeke** [comment]
@eli-schwartz I was wondering if this is on the roadmap somehow? Would help a lot with projects like [OpenBLAS](https://github.com/OpenMathLib/OpenBLAS) which do a lot of compilation on the same file with different flags.

For now my best bet is to define an array and traverse it after generating copies using configuration_file / generators.

**Reply by eli-schwartz** [reply]
I'm not certain I understand why copies are needed. Can you point me to what OpenBLAS is doing, exactly?

My understanding is that a cmake "object library" is exactly like a regular library except it doesn't create a static archive. So you could define an array and traverse it, generating one internal static library per set of flags, without doing any copying.

It is absolutely possible to include a single source file in multiple targets.

**Reply by HaoZeke** [reply]
Sorry, should have been clearer. Here is how the CMakeLists implements this:

https://github.com/OpenMathLib/OpenBLAS/blob/780497303c9f05013ebde29602830bc1cd69a050/interface/CMakeLists.txt#L114-L117

```cmake
  GenerateNamedObjects("imax.c" "" "i*max" ${CBLAS_FLAG})
  GenerateNamedObjects("imax.c" "USE_ABS" "i*amax" ${CBLAS_FLAG})
  GenerateNamedObjects("imax.c" "USE_ABS;USE_MIN" "i*amin" ${CBLAS_FLAG})
  GenerateNamedObjects("imax.c" "USE_MIN" "i*min" ${CBLAS_FLAG})
...
add_library(interface OBJECT ${OPENBLAS_SRC})
```

While the Makefile is a lot simpler:
https://github.com/OpenMathLib/OpenBLAS/blob/780497303c9f05013ebde29602830bc1cd69a050/interface/Makefile#L724-L734

```make
isamax.$(SUFFIX) isamax.$(PSUFFIX) : imax.c
	$(CC) $(CFLAGS) -c -DUSE_ABS -UUSE_MIN $< -o $(@F)

idamax.$(SUFFIX) idamax.$(PSUFFIX) : imax.c
	$(CC) $(CFLAGS) -c -DUSE_ABS -UUSE_MIN $< -o $(@F)

iqamax.$(SUFFIX) iqamax.$(PSUFFIX) : imax.c
	$(CC) $(CFLAGS) -c -DUSE_ABS -UUSE_MIN $< -o $(@F)

icamax.$(SUFFIX) icamax.$(PSUFFIX) : imax.c
	$(CC) $(CFLAGS) -c -DUSE_ABS -UUSE_MIN $< -o $(@F)
```

So basically in each case the same source file is used, just lightly modified (via the preprocessor defines). Right now I have the following:

```meson
configurations = [
  # TODO: Also add defines for DOUBLE for d and the rest
  {'defs': [], 'name': 'imax', 'source': 'imax.c'},  # Original, no defines
  {'defs': ['-DUSE_ABS'], 'name': 'iamax', 'source': 'imax.c'},
  {'defs': ['-DUSE_ABS', '-DUSE_MIN'], 'name': 'iamin', 'source': 'imax.c'}, 
  {'defs': ['-DUSE_MIN'], 'name': 'imin', 'source': 'imax.c'},
]

_static_libs = []
# TODO: Maybe use a generator for this
foreach conf : configurations
  named_src = configure_file(
    input: conf['source'],
    output:  conf['name'] + '.c',
    copy: true
  )
  _static_libs += static_library(conf['name'],
                                 named_src,
                                 include_directories: _inc,
                                 c_args: conf['defs'])
endforeach
```

Which feels a bit weird, and was wondering if there's a better approach.

**Reply by eli-schwartz** [reply]
My straightforward assumption for how to do this would be:

```meson
configurations = [
  # TODO: Also add defines for DOUBLE for d and the rest
  {'defs': [], 'name': 'imax'},  # Original, no defines
  {'defs': ['-DUSE_ABS'], 'name': 'iamax'},
  {'defs': ['-DUSE_ABS', '-DUSE_MIN'], 'name': 'iamin'}, 
  {'defs': ['-DUSE_MIN'], 'name': 'imin'},
]

_static_libs = []
foreach conf: configurations
    _static_libs += static_library(
        conf['name'],
        'imax.c',
        include_directories:  _inc,
        c_args: conf['defs'],
    )
endforeach
```

**Reply by HaoZeke** [reply]
Thanks for the prompt turnaround, that was my first thought but then the object files are seemingly not correctly generated. With the `configure_file` approach:
```bash
❯ nm bbdir/interface/libimax.a

bbdir/interface/libimax.a.p/meson-generated_.._imax.c.o:
                 U ismax_k
0000000000000000 T NAME
❯ nm bbdir/interface/libiamax.a

bbdir/interface/libiamax.a.p/meson-generated_.._iamax.c.o:
                 U isamax_k
0000000000000000 T NAME
```

Without it (the suggested variant):

```bash
❯ nm bbdir/interface/libimax.a

bbdir/interface/libimax.a.p/imax.c.o:
                 U ismax_k
0000000000000000 T NAME
❯ nm bbdir/interface/libiamax.a

bbdir/interface/libiamax.a.p/imax.c.o:
                 U isamax_k
0000000000000000 T NAME
```

Where the object file is the same for both (`imax.c.o`) which I thought would cause conflicts later (I also tested extracting the object out of the library where I ran into conflicts real quick)..

**Reply by eli-schwartz** [reply]
When you say "seemingly not correctly generated", can you elaborate on what you mean? As far as I can tell they have exactly the symbols you expect?

> Where the object file is the same for both (`imax.c.o`) which I thought would cause conflicts later (I also tested extracting the object out of the library where I ran into conflicts real quick)..

Extracting it for what, doing what with it?

**Reply by HaoZeke** [reply]
> When you say "seemingly not correctly generated", can you elaborate on what you mean? As far as I can tell they have exactly the symbols you expect?
> 
> > Where the object file is the same for both (`imax.c.o`) which I thought would cause conflicts later (I also tested extracting the object out of the library where I ran into conflicts real quick)..
> 
> Extracting it for what, doing what with it?

My bad, I was using it in:
```meson
_interface = static_library('_interface',
                    link_whole: _static_libs)
```

Which works alright both ways, so I'll stick to the simpler form you suggested, thanks!

P.S. By works right I mean the symbols in the library are correct, though the intermediates are a little oddly named:

```bash
# configure_file
bbdir/interface/libimax.a.p/meson-generated_.._imax.c.o:
                 U ismax_k
0000000000000000 T NAME

bbdir/interface/libiamax.a.p/meson-generated_.._iamax.c.o:
                 U isamax_k
0000000000000000 T NAME

bbdir/interface/libiamin.a.p/meson-generated_.._iamin.c.o:
                 U isamin_k
0000000000000000 T NAME

bbdir/interface/libimin.a.p/meson-generated_.._imin.c.o:
                 U ismin_k
0000000000000000 T NAME

# without configure_file
nm bbdir/interface/lib_interface.a

bbdir/interface/libimax.a.p/imax.c.o:
                 U ismax_k
0000000000000000 T NAME

bbdir/interface/libiamax.a.p/imax.c.o:
                 U isamax_k
0000000000000000 T NAME

bbdir/interface/libiamin.a.p/imax.c.o:
                 U isamin_k
0000000000000000 T NAME

bbdir/interface/libimin.a.p/imax.c.o:
                 U ismin_k
0000000000000000 T NAME
```
